### PR TITLE
Fix replies not being forwarded to all places where replied-to status would be displayed

### PR DESCRIPTION
**Pull request by Gargron**
**Behaviour change:** Before, a public reply to a local status would be forwarded to the followers of the parent's account, and that's it. However, the replied-to status has obviously also travelled to places where people have reblogged/favourited it, replied to it, and to places of people mentioned in it. This means that replies would not reach all places where the replied-to status was displayed. Now, we include all the places the replied-to status would go in the decision where the reply should go (and its delete, likewise).

___

Activity forwarding on Create and Delete handlers was using manual logic in separate places for selecting inboxes to deliver to. With this change, the central place to look where a status gets delivered, be it a local or a remote one, is now the StatusReachFinder (there is one more place in ActivityPub::DistributionWorker where StatusReachFinder is not used, but that change is implemented in #16697). New unit tests help keep the behaviour consistent.

Processing of `Collection-Synchronization` header is moved to the worker. Selected headers may be passed all the way down to the activity handlers if extra information from the delivery request is required.

### CODE REVIEW

**Comment by ClearlyClaire** [comment]
Do we actually want that? While I understand wishing to avoid partial conversations, I'm afraid this could generate some substantial workload on many instances

**Comment by Gargron** [comment]
> Do we actually want that? While I understand wishing to avoid partial conversations, I'm afraid this could generate some substantial workload on many instances

Surely the parent’s followers generate most of that load in the majority of cases

**Comment by ClearlyClaire** [comment]
> > Do we actually want that? While I understand wishing to avoid partial conversations, I'm afraid this could generate some substantial workload on many instances
> 
> Surely the parent’s followers generate most of that load in the majority of cases

On the contrary, I'd expect boosts of trending toots to do a lot here?

**Comment by Gargron** [comment]
> > > Do we actually want that? While I understand wishing to avoid partial conversations, I'm afraid this could generate some substantial workload on many instances
> > 
> > 
> > Surely the parent’s followers generate most of that load in the majority of cases
> 
> On the contrary, I'd expect boosts of trending toots to do a lot here?

I know of a few accounts with 8K followers (that aren't admins) but 8K boosts on any single status is a rarity yet. So that's my logic.

File: app/workers/activitypub/followers_synchronization_worker.rb [hunk 1]
```diff
@@ -6,9 +6,8 @@ class ActivityPub::FollowersSynchronizationWorker
   sidekiq_options queue: 'push', lock: :until_executed
 
   def perform(account_id, url)
-    @account = Account.find_by(id: account_id)
-    return true if @account.nil?
-
-    ActivityPub::SynchronizeFollowersService.new.call(@account, url)
+    ActivityPub::SynchronizeFollowersService.new.call(Account.find(account_id), url)
+  rescue ActiveRecord::RecordNotFound
+    true
```
Comment (ClearlyClaire) [inline app/workers/activitypub/followers_synchronization_worker.rb:11]: Not a fan of catching errors more broadly than where we expect them.

File: app/controllers/activitypub/inboxes_controller.rb [hunk 2]
```diff
@@ -57,20 +60,16 @@ def upgrade_account
...
-  rescue Parslet::ParseFailed
-    Rails.logger.warn 'Error parsing Collection-Synchronization header'
+  def process_payload
+    ActivityPub::ProcessingWorker.perform_async(
+      signed_request_account.id,
+      body,
+      @account&.id,
+      accepted_headers_from_request
+    )
   end
```
Comment (ClearlyClaire) [inline app/controllers/activitypub/inboxes_controller.rb:70]: I am honestly unsure why this is being moved elsewhere.

File: app/lib/activitypub/activity/create.rb [hunk 3]
```diff
@@ -490,7 +490,16 @@ def tombstone_exists?
   def forward_for_reply
     return unless @status.distributable? && @json['signature'].present? && reply_to_local?
 
-    ActivityPub::RawDistributionWorker.perform_async(Oj.dump(@json), replied_to_status.account_id, [@account.preferred_inbox_url])
+    # If the status is replying to a local status, we should forward it
+    # everywhere where that local status was distributed
+
+    status_reach_finder = StatusReachFinder.new(@status)
```
Comment (ClearlyClaire) [inline app/lib/activitypub/activity/create.rb:496]: Am I wrong in understanding that the computed reach includes, for instance, people mentioned by this status? Meaning you'd re-distribute to remote users who were explicitly mentioned by the remote author?

**Pull request closed without merging**